web/src/main/java/org/springframework/security/web/csrf/SecureCsrfToken.java

```diff
@@ -0,0 +1,151 @@
+
+	private final String parameterName;
+
+	private final String headerName;
+
+	private final SecureRandom secureRandom;
+
+	private final int randomByteSize;
+
+	/**
+	 * Creates a new instance
+	 *
+	 * @param headerName    the HTTP header name to use
+	 * @param parameterName the HTTP parameter name to use
+	 * @param token         the value of the token (i.e. expected value of the HTTP
+	 *                      parameter of parametername).
+	 */
+	public SecureCsrfToken(String headerName, String parameterName, String token) {
+		Assert.hasLength(headerName, "headerName cannot be null or empty");
+		Assert.hasLength(parameterName, "parameterName cannot be null or empty");
+		Assert.hasLength(token, "token cannot be null or empty");
+		this.headerName = headerName;
+		this.parameterName = parameterName;
+		this.tokenBytes = Utf8.encode(token);
+		this.secureRandom = new SecureRandom();
+		this.randomByteSize = this.tokenBytes.length;
+	}
+
+	/*
+	 * (non-Javadoc)
+	 *
+	 * @see org.springframework.security.web.csrf.CsrfToken#getHeaderName()
+	 */
+	public String getHeaderName() {
+		return this.headerName;
+	}
+
+	/*
+	 * (non-Javadoc)
+	 *
+	 * @see org.springframework.security.web.csrf.CsrfToken#getParameterName()
+	 */
+	public String getParameterName() {
+		return this.parameterName;
+	}
+
+	/*
+	 * (non-Javadoc)
+	 *
+	 * @see org.springframework.security.web.csrf.CsrfToken#getToken()
+	 */
+	public String getToken() {
+		ByteBuffer byteBuffer = ByteBuffer.allocate(this.randomByteSize);
+		this.secureRandom.ints(Byte.MIN_VALUE, Byte.MAX_VALUE).limit(randomByteSize)
+				.forEach((randInt) -> byteBuffer.put((byte) randInt));
+
+		byte[] randomBytes = byteBuffer.array();
+
+		byte[] xoredCsrf = xorCsrf(randomBytes, this.tokenBytes);
+
+		ByteBuffer combinedBuffer = ByteBuffer.allocate(this.randomByteSize + xoredCsrf.length);
+		combinedBuffer.put(randomBytes);
+		combinedBuffer.put(xoredCsrf);
+
+		// returning randomBytes + XOR csrf token
+		return Base64.getEncoder().encodeToString(combinedBuffer.array());
+	}
+
+	private static byte[] xorCsrf(byte[] randomBytes, byte[] csrfBytes) {
+		byte[] xoredCsrf = new byte[csrfBytes.length];
+		System.arraycopy(csrfBytes, 0, xoredCsrf, 0, csrfBytes.length);
+		for (byte b : randomBytes) {
+			for (int i = 0; i < xoredCsrf.length; i++) {
+				xoredCsrf[i] ^= b;
+			}
+		}
+
+		return xoredCsrf;
+	}
+
+	@Override
+	public boolean matches(String token) {
+		byte[] paramToken = null;
+
+		try {
+			paramToken = Base64.getDecoder().decode(token);
+		} catch (Exception ex) {
+			return false;
+		}
+
+		int tokenSize = this.tokenBytes.length;
+
+		if (paramToken.length == tokenSize) {
+			return MessageDigest.isEqual(this.tokenBytes, paramToken);
+		} else if (paramToken.length < tokenSize) {
+			return false;
+		}
+
+		// extract token and random bytes
+		int paramXorTokenOffset = paramToken.length - tokenSize;
+		ByteBuffer paramXoredToken = ByteBuffer.allocate(tokenSize);
```

Comment:
Again, I'd prefer to avoid the use of `ByteBuffer` here if possible

Reply:
will use `System.arraycopy` instead for extraction